src/nti/site/testing.py

```diff
@@ -0,0 +1,354 @@
+        """
+        install_main_application_and_sites(conn,
+                                           main_alias=self.main_application_folder_name,
+                                           main_setup=self.on_application_and_sites_installed)
+
+    def on_connection_opened(self, conn):
+        super(current_db_site_trans, self).on_connection_opened(conn)
+        main_name = hostpolicy.text_type(self.main_application_folder_name)
+
+        root = conn.root()
+        if main_name not in root:
+            self.on_main_application_folder_missing(conn)
+
+        sitemanc = conn.root()[hostpolicy.text_type(self.main_application_folder_name)]
+        if self._site_name:
+            with currentSite(sitemanc):
+                sitemanc = get_site_for_site_names((self._site_name,))
+
+        self._site_cm = currentSite(sitemanc)
+        self._site_cm.__enter__() # pylint:disable=no-member
+        assert component.getSiteManager() == sitemanc.getSiteManager()
+        return conn
+
+    def __exit__(self, t, v, tb):
+        result = self._site_cm.__exit__(t, v, tb) # pylint:disable=no-member
+        super(current_db_site_trans, self).__exit__(t, v, tb)
+        return result
+
+
+mock_db_trans = current_db_site_trans # BWC, remove in 2021
+reset_db_caches = zodb.reset_db_caches # BWC, remove in 2021
+
+def _mock_ds_wrapper_for(func, installer_factory, installer_kwargs, db_factory,
+                         marker=object()):
+
+    @functools.wraps(func)
+    def f(*args):
+        # we may not be in a layer that's done this. Note that we don't tear it down though.
+        # This is needed to run ``install_main_application_and_sites``
+        setHooks()
+        db = db_factory()
+
+        old_db = zodb.ZODBLayer.db
+        old_func_db = marker
+        try:
+            zodb.ZODBLayer.db = db
+            if SharedConfiguringTestLayer.current_test is not None:
+                SharedConfiguringTestLayer.current_test.db = db
+            if args and isinstance(args[0], unittest.TestCase):
+                # self
+                old_func_db = getattr(args[0], 'db', marker)
+                args[0].db = db
+
+            with installer_factory(db, **installer_kwargs):
+                pass
+
+            sitemanc = SiteManagerContainer()
+            sitemanc.setSiteManager(BTreeLocalSiteManager(None))
+
+            with currentSite(sitemanc):
+                assert component.getSiteManager() == sitemanc.getSiteManager()
+                func(*args)
+        finally:
+            if args and isinstance(args[0], unittest.TestCase):
+                if old_func_db is marker:
+                    del args[0].db
+                else:
+                    args[0].db = old_func_db
+            db.close()
+            zodb.ZODBLayer.db = old_db
+    return f
+
+
+def default_db_factory():
+    return ZODB.DB(DemoStorage(name='Users'))
+
+def uses_independent_db_site(*args, **kwargs):
```

Comment:
This was the version of `WithMockDSTrans` that was here. It's interesting to note that this, and the other copies of it, created the DB object when the decorator runs, i.e., at import time. That makes imports more expensive than they need to be and contributes memory usage since it can never be collected. It's now deferred until the test actually runs.